datafusion/src/physical_plan/expressions/binary.rs

```diff
@@ -1381,4 +1416,37 @@ mod tests {
             ))
         }
     }
+
+    #[test]
+    fn relatively_deeply_nested() {
+        // Reproducer for https://github.com/apache/arrow-datafusion/issues/419
+
+        // where even relatively shallow binary expressions overflowed
+        // the stack in debug builds
+
+        let input: Vec<_> = vec![1, 2, 3, 4, 5].into_iter().map(Some).collect();
+        let a: Int32Array = input.iter().collect();
+
+        let batch = RecordBatch::try_from_iter(vec![("a", Arc::new(a) as _)]).unwrap();
+        let schema = batch.schema();
+
+        // build a left deep tree ((((a + a) + a) + a ....
+        let tree_depth: i32 = 100;
```

Comment:
On master, this test causes a stack overflow with `tree_depth` of `10`. After the changes in this PR it passes successfully with a tree_depth of 100 (I didn't try any bigger)

Reply:
🎉